### PR TITLE
More secure twitter fix

### PR DESCRIPTION
**Pull request by jhung**
Using oAuth Twitter for Developers plugin for a more secure twitter feed without API keys being made public in the PHP file.

To test:
1. Install the [oAuth Twitter Feed for Developers](https://wordpress.org/plugins/oauth-twitter-feed-for-developers/) plugin.
2. In the WP dashboard > Settings > Twitter Feed Auth, add the Twitter API keys for the IDI Site (can be found on dev.twitter.com when logged in as FluidProject). Save the changes.
3. merge the changes to the IDI-theme
4. reload IDI site and look at the index page

The old keys which were previously public are no longer valid. New keys have been generated.

### CODE REVIEW

File: functions.php [hunk 1]
```diff
@@ -120,40 +120,77 @@ function panel_login_fail( $username ) {
     if ( !empty($referrer) && !strstr($referrer,'wp-login') && !strstr($referrer,'wp-admin') ) {
         $referrer = explode('?', $referrer);
         parse_str($referrer[1], $referrer_args);
-        unset($referrer_args['idilogout']); 	//unset old idilogout arg if set
-        $referrer_args['idilogin'] = "failed"; 	//let theme know there are errors
+        unset($referrer_args['idilogout']);     //unset old idilogout arg if set
+        $referrer_args['idilogin'] = "failed";     //let theme know there are errors
```
Comment (cindyli) [inline functions.php]: This code has a mixed use of single quotes (') and double quotes ("). Can you check the coding convention for wordpress plugins and unify the usage? Thanks.

Comment (jhung) [inline functions.php]: There's no strict convention. Their [styleguide](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/) says:

"Use single and double quotes when appropriate. If you’re not evaluating anything in the string, use single quotes."

In this case I'll change it to single quotes to be consistent with the usage so far.

File: functions.php [hunk 2]
```diff
@@ -120,40 +120,77 @@ function panel_login_fail( $username ) {
...
+            $the_tweet = $tweet['text'];
+            /*
+            Format the Twitter information for output. This is a modified version of the source script. It's been changed to match our classnames and layout.
+            Source: https://github.com/stormuk/storm-twitter-for-wordpress/wiki/Example-code-to-layout-tweets
+
+
+            Convert Tweet Entities within the Tweet text to link to their URLs on Twitter. */
+
+            // Convert User_mentions to link to user profiles.
+            if(is_array($tweet['entities']['user_mentions'])){
```
Comment (cindyli) [inline functions.php]: These 3 section of code for converting `user_mentions`/`hashtags`/`entities` are quite similar. Please refactor the code to use a common function for it. This function could have 4 parameters:

1. `the_tweet`;
2. The field name. The value in each case would be `user_mentions`, `hashtags`, `entities`;
3. `pattern`. The first argument of `preg_replace()`;
4. `replacement`. The second argument of `preg_replace()`.

File: functions.php [hunk 3]
```diff
@@ -120,40 +120,77 @@ function panel_login_fail( $username ) {
...
-					<a class="twitter-follow-button" rel="external nofollow" href="http://twitter.com/'.$twitter_un.'" title="Follow @'.$twitter_un.'">@'.$twitter_un.'</a>
-					<ul>
-	';
-	if(!empty($tweets)){
-	        foreach($tweets as $tweet) {
-                echo '<li class="tweet">'.$tweet->text.
-                '<div class="tweet-date">'.substr($tweet->created_at, 0, 16).'</div></li>';
+    $tweets = getTweets($num_tweets, $twitter_un);
+
+    echo '<div class="idi-box idi-highlight-box twitter-feed-group">
```
Comment (cindyli) [inline functions.php]: `echo` will output wrong indentations for the html. The first line would have no indent while the following line would start with much more indents. Please consolidate. Thanks.

This comment applies to all other `echo` commands.

File: functions.php (hunk 2 above)
Comment (jhung) [inline functions.php]: I looked into this previously. Although the logic is the same in 3 cases, the input and output are all different in each case (each loop uses different arrays, and each loop outputs a different string).

Maybe there's a way to generalize this, but I couldn't think of a way to do this effectively. Any suggestions?

Comment (cindyli) [inline functions.php]: This is my implementation of the function and how to call it to deal with the case of `user_mentions`:

```
function wireUpLinks($tweet, $entitiesField, $fieldToMatch, $pattern, $replacement) {
    $the_tweet = $tweet['text'];
    $entity = $tweet['entities'][$entitiesField];
    if (is_array($entity)) {
        foreach ($entity as $key => $value) {
            $matchedValue = $value[$fieldToMatch];
            $pattern = str_replace('%replace', $matchedValue, $pattern);
            $replacement = str_replace('%replace', $matchedValue, $replacement);
            $the_tweet = preg_replace($pattern, $replacement, $the_tweet);
        }
    }

    return $the_tweet;
}

$pattern = '/@%replace/i';
$replacement = '<a href="http://www.twitter.com/%replace" target="_blank">@%replace</a>';
$output_tweet = wireUpLinks($tweet, 'hashtags', 'user_mentions', $pattern, $replacement);
```

This function should be called for other two cases too. Please feel free to improve it. Thanks.

**Comment by jhung** [comment]
The API has changed a little and I am now working on a new solution.

**Pull request closed without merging**